File: leetcode/39.combination-sum/go/combinationsum.go
package combinationsum

func combinationSum(candidates []int, target int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)
	sum := 0

	var backtrack func(start int)

	backtrack = func(start int) {
		if sum == target {
			tmp := make([]int, len(path))
			copy(tmp, path)
			res = append(res, tmp)
			return
		}

		for i := start; i < len(candidates); i++ {
			if sum+candidates[i] > target {
				continue
			}
			
			path = append(path, candidates[i])
			sum += candidates[i]
			backtrack(i)
			sum -= candidates[i]
			path = path[:len(path)-1]
		}
	}

	backtrack(0)
	return res
}

File: leetcode/39.combination-sum/ts/combinationsum.ts
const combinationSum = (candidates: number[], target: number): number[][] => {
    const res : number[][] = []
    const path : number[] = []
    let sum = 0

    const backtrack = (start: number) => {
        if (sum === target) {
            res.push([...path])
            return
        }

        for (let i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) continue

            path.push(candidates[i])
            sum += candidates[i]
            backtrack(i)
            sum -= candidates[i]
            path.pop()
        }
    }

    backtrack(0)
    return res
}
